File: leetcode/78_subsets.py
class Solution(object):
    def subsets(self, nums):
        res = []
        for i in range(1 << len(nums)):
            tmp = []
            for j in range(len(nums)):
                if i & 1 << j:
                    tmp.append(nums[j])
            res.append(tmp)
        return res


"""
0 1 2
000 & (001, 010, 100) => []
001 & (001, 010, 100) => [0]
010 & (001, 010, 100) => [1]
011 & (001, 010, 100) => [0, 1]
100 & (001, 010, 100) => [2]
101 & (001, 010, 100) => [0, 2]
110 & (001, 010, 100) => [1, 2]
111 & (001, 010, 100) => [0, 1, 2]
"""

File: leetcode/java/78Subsets.java
import java.util.ArrayList;
import java.util.List;

class Solution {
  private List<List<Integer>> result;
  private int[] numbers;

  public List<List<Integer>> subsets(int[] nums) {
    result = new ArrayList<>();
    numbers = nums;
    backtrack(new ArrayList<>(), 0);
    return result;
  }

  private void backtrack(List<Integer> subset, int index) {
    result.add(subset);
    for (int i = index; i < numbers.length; i++) {
      List<Integer> newSubset = new ArrayList<>(subset);
      newSubset.add(numbers[i]);
      backtrack(newSubset, i + 1);
    }
  }
}
